File: algorithms/2_medium/A/00870_advantage_shuffle/solution.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static vector<int> advantageCount(vector<int>& nums1, const vector<int>& nums2) {
        const isize len = std::ssize(nums1);
        assert(len == nums2.size() && len >= 1 && len <= 100'000);

        auto combo2 = vector<pair<int, int>>();
        combo2.reserve(len);
        for (isize i = 0; i < len; i += 1) {
            combo2.push_back(make_pair(nums2[i], static_cast<int>(i)));
        }

        auto res = vector<int>(len, 0);

        ranges::sort(nums1);
        ranges::sort(combo2);

        isize begin = 0;
        isize end = len;

        for (auto it = combo2.crbegin(); it != combo2.crend(); it += 1) {
            const auto [num, idx] = *it;

            if (num < nums1[end - 1]) {
                res[idx] = nums1[end - 1];
                end -= 1;
            } else {
                res[idx] = nums1[begin];
                begin += 1;
            }
        }

        return res;
    }
};

File: algorithms/2_medium/A/00870_advantage_shuffle/solution.rs
impl Solution {
    pub fn advantage_count(mut nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let len = nums1.len();
        assert!(len == nums2.len() && len >= 1 && len <= 100_000);

        let mut combo2: Vec<(i32, u32)> = nums2.into_iter().enumerate().map(|(i, v)| (v, i as u32)).collect();
        let mut res = vec![0_i32; len];

        nums1.sort_unstable();
        combo2.sort_unstable();

        let (mut begin, mut end) = (0_usize, len);

        for (num, idx) in combo2.into_iter().rev() {
            if num < nums1[end - 1] {
                res[idx as usize] = nums1[end - 1];
                end -= 1;
            } else {
                res[idx as usize] = nums1[begin];
                begin += 1;
            }
        }

        res
    }
}
